doc/sr_sarra.8.rst

=========
 SR_Sarra
=========

---------------------------------------------------------
Subscribe, Acquire and Recursively Re-announce Ad nauseam
---------------------------------------------------------

:Manual section: 8
:Date: @Date@
:Version: @Version@
:Manual group: MetPx Sarracenia Suite

.. contents::


SYNOPSIS
========

**sr_sarra** foreground|start|stop|restart|reload|status configfile

**sr_sarra** cleanup|declare|setup configfile


DESCRIPTION
===========

**sr_sarra** is a program that Subscribes to file notifications,
Acquires the files and ReAnnounces them at their new locations.
The notification protocol is defined here `sr_post(7) <sr_post.7.rst>`_

**sr_sarra** connects to a *broker* (often the same as the remote file server
itself) and subscribes to the notifications of interest. It uses the notification
information to download the file on the local server it's running on.
It then posts a notification for the downloaded files on a broker (usually on the local server).

**sr_sarra** can be used to acquire files from `sr_post(1) <sr_post.1.rst>`_
or `sr_watch(1) <sr_watch.1.rst>`_  or to reproduce a web-accessible folders (WAF),
that announce its products.

The **sr_sarra** is an `sr_subscribe(1) <sr_subscribe.1.rst>`_  with the following presets::

   mirror True


Specific consuming requirements
--------------------------------

If the messages are posted directly from a source,
the exchange used is 'xs_<brokerSourceUsername>'.
Such message may not contain a source nor an origin cluster,
or a malicious user may set the values incorrectly.
To protect against malicious settings, administrators
should set *source_from_exchange* to **True**.


- **source_from_exchange  <boolean> (default: False)**

Upon reception, the program will set these values
in the parent class (here cluster is the value of
option **cluster** taken from default.conf):

self.msg.headers['source']       = <brokerUser>
self.msg.headers['from_cluster'] = cluster

overriding any values present in the message.  This setting
should always be used when ingesting data from a 
user exchange.


SEE ALSO
========

`sr_subscribe(1) <sr_subscribe.1.rst>`_ - the format of configurations for MetPX-Sarracenia.

`sr_report(7) <sr_report.7.rst>`_ - the format of report messages.

`sr_report(1) <sr_report.1.rst>`_ - process report messages.

`sr_post(1) <sr_post.1.rst>`_ - post announcemensts of specific files.

`sr_post(7) <sr_post.7.rst>`_ - the format of announcements.

`sr_subscribe(1) <sr_subscribe.1.rst>`_ - the download client.

`sr_watch(1) <sr_watch.1.rst>`_ - the directory watching daemon.

`dd_subscribe(1) <dd_subscribe.1.rst>`_ - the http-only download client.


CHANGES.rst
===============
 Release Notes 
===============

lists all changes between versions.

**2.18.07b3**

* bugfix     regression for subprocess api change @ V3.5
*            

**2.18.07b2**

* bugfix      HPC mirroring reports crash in retry logic, work-around, 
*             add try/except around it.
*             retry_ttl unit conversion error (was comparing milliseconds to seconds.) 
* new         poll_email_fetch - query and download from mailboxes. From Marie!
*

**2.18.07b1**

* bugfix      sr_audit now fixes missing instances. (issue #62 & #63 )
*             more output of plugin programs present in logs (issue #63)
*             two different crashes fixed in flow_test.sh
*             filter_wmo2msc directory tree naming improved ( issue #60 )
*             many documentation improvements. (Alain & Marie)
* new         enhanced parsing of date substitution (issue #55 Wahaj!)
*             now have program settings audit.conf  (issue #64 )


**2.18.06b2**

*             sr_poll bug for polling scripts fix from Jun.

**2.18.06b1**

* bugfix      sr_audit now runs for all users (restarts crashed components.)
* new         start post (sleep <= 0) now does nothing.
*             tested and added build instructions for RPM systems.
* doc         French docs done.
*             *sci-fi* (future planned features) removed from docs.
*             website migrated from sf.net to git repo ones (github, gitlab)

**2.18.05b4**

* bugfix      plugin msg_filter_wmo2msc.py fixed
*             sr_poll now uses its own set_dir_pattern to replace variables
*             sr_post/sr_poll/sr_watch fixed that a cache file for instance 000 was created
*             sr_post/sr_poll/sr_watch add caching of 'L' message
* new         sr_amqp for consumer sets channel.basic_recover(requeue=True)
* update      flow_post.sh posts lot of files within one call (load problem fixed)
*             flow_test does not test filename with space anymore
*             flow_test 5 instances for sender
* doc         miscellaneous doc fixups.

**2.18.05b2**

* bugfix       sr_sftp use chdir to see if connection is still alive
*              sr_http some site do not tolerate '//' in path
*              sr_subscribe some code added for sr_sender (lost in inheritance)
*              sr_file if a file does not exist and should be copied/linked just warn (Dominic)
* update       flow_test test2_f61.conf no use of post_total_save (race condition)

**2.18.05b1**

* bugfix       sr_sftp  differentiate put part vs put file 
* new          sr_retry uses caching for message uniqueness
*              sr_ftp,sr_sftp better connection test
* update       sr_poll  default post_base_url from destination url without password
*              sr_subscribe,sr_sender  log attempt
*              sr_subscribe retry.on_heartbeat on startup
*              sr_sender/sr_util if file to send does not exist... give it "attempt" chances
* doc          fixes and translation into french in progress
*              updated examples link, and samples directory cleanup

**2.18.04b5**

* bugfix  open local file read rb (instead of r+b)
*         remove .tmp file if upload does not work
*         sftp link and directory removal (from message)
* update  instance string using 2 digits (and code to migrate to)
*         switch for subprocess.check_call or subprocess.run 
*         give all chances to sigterm to complete
* new     inflight tmp/
*         If things go badly (general exception catch) keep message in retry
*         log when a retry message is dropped because it expires
*         poll accepts https (more test to do...not working with japan)
* doc     expanded description of *expire* setting.
*         Adding hint about use of *expire* option, how it is necessary for operations.
*         default setting is to avoid broker overload, too low for operational use.

**2.18.04b4**

* bugfix  destfn_script was not working
* update  sr_config.run_command presents subprocess check_call or run depending on python version
* new     inflight tmp/

**2.18.04b3**

* testing flow_test standardisation of messages
* bugfix  hb_retry no more dependency with on_start
*         sr_poll make sure all comparaisons are done without of trailing \n
* new     registered do_get,do_put
*         registered plugin, if return None, let python do it
*         plugin accel_scp.py
*         C has its own sarrac git repo
*         sr_config.py  '\ ' backslash_space allowed in options 
*         sr_config.py/sr_message.py  topic encodes  ' ' '#' with %20 %23
* update  flow_test standardisation of messages
*         sr_poll logs a warning when sleep time makes no sense
*         documentation launchpat for sundew
*         sr_audit log a message when --users all is done
*                   make sure heartbeat is in try/except 

**2.18.04b2**

* testing: Added recovery of flow_test stuck in retrying state.
* bugfix  C Truncate all headers and topic so they don't exceed 255 (AMQP limit.)
*         C Try to avoid being in conflict with stdin/out/err  open/close + 2 dup
*         C Valgrind hygiene: if nanosecond timestamp was 0, weird stuff happenned.  Fixed.
*           now valgrind does not complain at all.
*         C libsrshim enforced checks on commands'status
*         C any Python, topic and path with # encoded into %23 (as blank into %20)
* update  sr_audit hb_police_queues to check queue as admin, 
*           hb_sanity to check processes and sanity_log_dead option added            
*           no sleep option, sleep computed to trigger next heartbeat
*         sr_rabbit rabbit dependant commands placed in this file
* new     plugin do_send_log

**2.18.04b1**

* testing:retries on python side to validate products and routing
*         flow_post: loop on sorted individial products (spaces in path)... symlinks considered
* new     sr_subscribe: traceback logs when doing badly
*         sr_audit heartbeat works ... needs a config in audit/x.conf for now.

sarra/sr_sarra.py
#!/usr/bin/env python3
#
# This file is part of sarracenia.
# The sarracenia suite is Free and is proudly provided by the Government of Canada
# Copyright (C) Her Majesty The Queen in Right of Canada, Environment Canada, 2008-2015
#
# Questions or bugs report: dps-client@ec.gc.ca
# sarracenia repository: git://git.code.sf.net/p/metpx/git
# Documentation: http://metpx.sourceforge.net/#SarraDocumentation
#
# sr_sarra.py : python3 program allowing users to listen and download product from
#               another sarracenia server or from user posting (sr_post/sr_watch)
#               and reannounce the product on the current server
#
# Code contributed by:
#  Michel Grenier - Shared Services Canada
#  Last Changed   : Mon Sep 25 20:45 UTC 2017
#                   code rewritten : sr_sarra is an instantiation of sr_subscribe
#
########################################################################
#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful, 
#  but WITHOUT ANY WARRANTY; without even the implied warranty of 
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the 
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307  USA
#
#============================================================
# usage example
#
# sr_sarra [options] [config] [foreground|start|stop|restart|reload|status|cleanup|setup]
#
# sr_sarra consumes message, for each message it downloads the product
# and reannounce it. On usage of sarra is to acquire data from source
# that announce their products.  The other usage is to dessiminate products
# from other brokers/pumps.
#
# condition 1: from a source
# broker                  = where sarra is running (manager)
# exchange                = xs_source_user
# product                 = downloaded under directory (option document_root)
#                         = subdirectory from mirror option  OR
#                           message.headers['rename'] ...
#                           can be trimmed by option  strip
#
# post_broker             = where sarra is running (manager)
# post_exchange           = xpublic
# post_message            = same as incoming message
#                           message.headers['source']  is set from source_user
#                           message.headers['cluster'] is set from option cluster from default.conf
#                           message is built from url option give
#
# report_exchange         = xreport
#
#
# condition 2: from another broker/pump
# broker                  = the remote broker...
# exchange                = xpublic
# product                 = usually the product placement is mirrored 
#                           option document_root needs to be set
# post_broker             = where sarra is running (manager)
# post_exchange           = xpublic
# post_message            = same as incoming message
#                           message.headers['source']  left as is
#                           message.headers['cluster'] left as is 
#                           option url : gives new url announcement for this product
# report_exchange         = xs_"remoteBrokerUser"
#
#
#============================================================

import os,sys,time

try :    
         from sr_subscribe      import *
         print( "Using local module definitions, not system ones")
except : 
         from sarra.sr_subscribe import *

class sr_sarra(sr_subscribe):

    def check(self):

        if self.config_name == None : return

        if self.broker == None :
           self.logger.error("no broker given")
           self.help()
           sys.exit(1)

        if self.post_broker == None : self.post_broker = self.broker

        # exchanges suffix process if needed

        if self.exchange == None and self.exchange_suffix :
           self.exchange = 'xs_%s' % self.broker.username + self.exchange_suffix

        if self.post_exchange == None and self.post_exchange_suffix :
           self.post_exchange = 'xs_%s' % self.post_broker.username + self.post_exchange_suffix

        # verify exchange

        if self.exchange == None :
           self.logger.error("no exchange given")
           self.help()
           sys.exit(1)

        # verify post_base_dir

        if self.post_base_dir == None :
           if self.post_document_root != None :
              self.post_base_dir = self.post_document_root
              self.logger.warning("use post_base_dir instead of post_document_root")
           elif self.document_root != None :
              self.post_base_dir = self.document_root
              self.logger.warning("use post_base_dir instead of document_root")

        # bindings should be defined 

        if self.bindings == []  :
           key = self.topic_prefix + '.#'
           self.bindings.append( (self.exchange,key) )
           self.logger.debug("*** BINDINGS %s"% self.bindings)

        # default queue name if not given

        if self.queue_name == None :
           self.queue_name  = 'q_' + self.broker.username + '.'
           self.queue_name += self.program_name + '.' + self.config_name 

        # ===========================================================
        # some sr_subscribe options reset to match sr_sarra behavior
        # ===========================================================

        # currentDir is post_document_root if unset

        if self.currentDir == None :
           self.currentDir = self.post_document_root

        # always download ...

        if self.notify_only :
           self.logger.error("sarra notify_only True")
           os._exit(1)

        # we dont save nor restore

        if self.save or self.restore :
           self.logger.error("sarra no save/restore support")
           sys.exit(1)

        # we dont discard

        if self.discard :
           self.logger.error("sarra discard True")
           sys.exit(1)

        # retry_ttl setup.
        if self.retry_ttl == None:
           self.retry_ttl = self.expire

        if self.retry_ttl == 0:
           self.retry_ttl = None

        if self.retry_mode :
           self.execfile("plugin",'hb_retry')


        # default reportback if unset

        if self.reportback == None : self.reportback = False

        # do_task should have doit_download for now... make it a plugin later
        # and the download is the first thing that should be done

        if not self.doit_download in self.do_task_list :
           self.do_task_list.insert(0,self.doit_download)

        # MG FIXME : I dont think I forgot anything but if some options need
        #            to be specifically set for sr_sarra put them HERE

    def overwrite_defaults(self):

        # overwrite defaults
        # the default settings in most cases :
        # sarra receives directly from sources  onto itself
        # or it consumes message from another pump
        # we cannot define a default broker exchange

        # since it can be xreport or xs_remotepumpUsername ?
        # default broker and exchange None

        # in most cases, sarra downloads and repost for itself.
        self.inflight = None

        # default post_broker and post_exchange are

        self.post_exchange  = 'xpublic'
        if hasattr(self,'manager'):
           self.post_broker = self.manager

        # most of the time we want to mirror product directory and share queue

        self.mirror         = True

        # no directory if not provided

        self.currentDir     = None


        # ===========================================================
        # some sr_subscribe options reset to understand user sr_sarra setup
        # ===========================================================

        self.discard     = False
        self.notify_only = False
        self.restore     = False
        self.save        = False

        self.reportback  = None

        self.accept_unmatch = True



# ===================================
# MAIN
# ===================================

def main():

    args,action,config,old = startup_args(sys.argv)

    sarra = sr_sarra(config,args,action)
    sarra.exec_action(action,old)

    os._exit(0)

# =========================================
# direct invocation
# =========================================

if __name__=="__main__":
   main()

